piksi_tools/fileio.py

```diff
@@ -10,209 +11,416 @@
 # WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
 
 from __future__ import absolute_import, print_function
-from future.builtins import bytes  # makes python 2 bytes more similar to python 3
+from future.builtins import bytes  # makes python 2 `bytes()` more similar to python 3
+
+from collections import defaultdict
 
-import copy
 import itertools
 import random
 import time
 import threading
 import sys
 
+from six.moves.queue import Queue
+
 from sbp.client import Framer, Handler
-from sbp.file_io import (SBP_MSG_FILEIO_READ_DIR_RESP,
-                         SBP_MSG_FILEIO_READ_RESP, SBP_MSG_FILEIO_WRITE_RESP,
+from sbp.file_io import (SBP_MSG_FILEIO_READ_DIR_RESP, SBP_MSG_FILEIO_READ_RESP,
+                         SBP_MSG_FILEIO_WRITE_RESP, SBP_MSG_FILEIO_CONFIG_RESP,
                          MsgFileioReadDirReq, MsgFileioReadDirResp,
-                         MsgFileioReadReq, MsgFileioRemove, MsgFileioWriteReq)
+                         MsgFileioReadReq, MsgFileioRemove, MsgFileioWriteReq,
+                         MsgFileioConfigReq, MsgFileioConfigResp)
 
 from piksi_tools import serial_link
 
 MAX_PAYLOAD_SIZE = 255
 SBP_FILEIO_WINDOW_SIZE = 100
-SBP_FILEIO_TIMEOUT = 5.0
-MINIMUM_RETRIES = 3
+SBP_FILEIO_BATCH_SIZE = 1
+SBP_FILEIO_TIMEOUT = 3
+MAXIMUM_RETRIES = 20
 PROGRESS_CB_REDUCTION_FACTOR = 100
```

Comment:
Should PROGRESS_CB_REDUCTION_FACTOR change based upon the rate at which FILEIO is operating?  I tuned this based upon the speed of the old file transfer, but now if we move 3 times faster, I think perhaps this factor should be reduced accordingly when we receive the config response.  Basically I saw that updating the GUI was eating CPU and making a crash of the console more likely.

Reply:
ok, I can experiment with that

Reply:
Added some code to double PROGRESS_CB_REDUCTION_FACTOR if we're in the "new" config

Reply:
I'm still seeing a lot of "We dropped a packet, skipping this observation sequence..." logs, but I am not convinced it's the console that's dropping things... I think in this high bandwidth scenario we might be dropping data that's coming out of rpmsg somehow.  The dropped packet logging gets much worse if I up the solution rate.

Reply:
Also, check out e1139f4c6d9d8deea60b3f587906995ee1e3b5f0 -- which adjust the logic around this message, before this change I was seeing that we always logged twice when we got into this situation.